### PR TITLE
IF: Add a new set_finalizer (with weights, threshold, and local finalizer) to libtester and add more tests using it

### PR DESCRIPTION
**Pull request by linh2931**
* Add a new set_finalizer which takes  threshold, weight and local finalizers as arguments
* Using the new function, add tests `threshold_equal_to_half_weight_sum_test`, `votes_equal_to_threshold_test`, `votes_greater_than_threshold_test`, and `votes_less_than_threshold_test`.

Resolved https://github.com/AntelopeIO/leap/issues/2193

### CODE REVIEW

File: unittests/api_tests.cpp [hunk 1]
```diff
@@ -3917,4 +3917,118 @@ BOOST_AUTO_TEST_CASE(set_finalizer_test) { try {
    BOOST_CHECK_GT(lib, lib_after_transition);
 } FC_LOG_AND_RETHROW() }
 
+void test_finality_transition(const vector<account_name> accounts, const base_tester::finalizer_policy_input& input, bool lib_advancing_expected) {
```
Comment (heifner) [inline unittests/api_tests.cpp:3920]: Can be `const&`

**Review by heifner — CHANGES_REQUESTED** [review]

File: unittests/api_tests.cpp [hunk 2]
```diff
@@ -3917,4 +3917,118 @@ BOOST_AUTO_TEST_CASE(set_finalizer_test) { try {
...
+   // Carol votes with weight 5 and threshold 5
+   test_finality_transition(account_names, ploicy_input, true); // lib_advancing_expected
+} FC_LOG_AND_RETHROW() }
+
+BOOST_AUTO_TEST_CASE(votes_greater_than_threshold_test) { try {
+   vector<account_name> account_names = {
+      "alice"_n, "bob"_n, "carol"_n
+   };
+
+   base_tester::finalizer_policy_input ploicy_input = {
```
Comment (greg7mdp) [inline unittests/api_tests.cpp:4005]: type in all the `ploicy_input`.

File: libraries/testing/include/eosio/testing/tester.hpp [hunk 3]
```diff
@@ -252,7 +252,24 @@ namespace eosio { namespace testing {
          transaction_trace_ptr       set_producers(const vector<account_name>& producer_names);
          transaction_trace_ptr       set_producer_schedule(const vector<producer_authority>& schedule);
          transaction_trace_ptr       set_producers_legacy(const vector<account_name>& producer_names);
-         transaction_trace_ptr       set_finalizers(const vector<account_name>& finalier_names);
+
+         // libtester uses 1 as weight of each of the finalizer, sets (2/3 finalizers + 1)
+         // as threshold, and makes all finalizers vote QC
+         transaction_trace_ptr  set_finalizers(const vector<account_name>& finalier_names);
```
Comment (greg7mdp) [inline libraries/testing/include/eosio/testing/tester.hpp:258]: typo in `finalier_names`

**Review by greg7mdp — APPROVED** [review]

File: libraries/testing/tester.cpp [hunk 4]
```diff
@@ -1215,6 +1215,36 @@ namespace eosio { namespace testing {
                           fc::mutable_variant_object()("finalizer_policy", std::move(fin_policy_variant)));
    }
 
+   transaction_trace_ptr base_tester::set_finalizers(const finalizer_policy_input& input ) {
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/testing/tester.cpp:1218]: Thanks. I will make the change.

File: libraries/testing/include/eosio/testing/tester.hpp (hunk 3 above)
Comment (linh2931) [reply libraries/testing/include/eosio/testing/tester.hpp:258]: Thanks!

File: unittests/api_tests.cpp (hunk 2 above)
Comment (linh2931) [reply unittests/api_tests.cpp:4005]: Will correct them.

**Review by heifner — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Updated set finalizer which takes threshold, weight and local finalizers as arguments.
Note:end

**Comment by ericpassmore** [comment]
Note:start
category: Tests
component: Internal
summary: Updated set finalizer which takes threshold, weight and local finalizers as arguments.
Note:end

**Pull request merged**